teos/src/main.rs

```diff
@@ -69,6 +69,10 @@ fn create_new_tower_keypair(db: &DBM) -> (SecretKey, PublicKey) {
 async fn main() {
     let opt = Opt::from_args();
     let path = config::data_dir_absolute_path(opt.data_dir.clone());
+    let conf_file_path = path.join("teos.toml");
+
+    // Log config file path
+    log::info!("Using configuration file: {:?}", conf_file_path);
```

Comment:
> > This is still being logged before the logger is initialized, hence it would never be shown
> 
> Tried to log it after logger but shows error `move occurs because conf_file_path has type pathbuf , which does implement the copy trait`

So, I did logged before the `config::from_file`

Reply:
Make it so `Config:: from_file` borrows a `Pathbuf` instead of taking ownership of it, then you'll be able to just pass a reference to `conf_file_path` instead of moving it.

Reply:
> 

I referenced it with `let mut conf = config::from_file::<Config>(&conf_file_path);` but still shows error and states `consider removing the borrow`.

Reply:
> > 
> 
> I referenced it with `let mut conf = config::from_file::<Config>(&conf_file_path);` but still shows error and states `consider removing the borrow`.

Like this:-

`async fn main() {
    let opt = Opt::from_args();
    let path = config::data_dir_absolute_path(opt.data_dir.clone());
    let conf_file_path = path.join("teos.toml");
    // Create data dir if it does not exist
    fs::create_dir_all(&path).unwrap_or_else(|e| {
        eprintln!("Cannot create data dir: {e:?}");
        std::process::exit(1);
    });

    // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
    let mut conf = config::from_file::<Config>(&conf_file_path);
    let is_default = conf.is_default();
    conf.patch_with_options(opt);
    conf.verify().unwrap_or_else(|e| {
        eprintln!("{e}");
        std::process::exit(1);
    });`

Reply:
> > > 
> > 
> > 
> > I referenced it with `let mut conf = config::from_file::<Config>(&conf_file_path);` but still shows error and states `consider removing the borrow`.
> 
> Like this:-
> 
> `async fn main() { let opt = Opt::from_args(); let path = config::data_dir_absolute_path(opt.data_dir.clone()); let conf_file_path = path.join("teos.toml"); // Create data dir if it does not exist fs::create_dir_all(&path).unwrap_or_else(|e| { eprintln!("Cannot create data dir: {e:?}"); std::process::exit(1); });
> 
> ```
> // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
> let mut conf = config::from_file::<Config>(&conf_file_path);
> let is_default = conf.is_default();
> conf.patch_with_options(opt);
> conf.verify().unwrap_or_else(|e| {
>     eprintln!("{e}");
>     std::process::exit(1);
> });`
> ```

Then after log I have added this:-

`
    // Set log level
    SimpleLogger::new()
        .with_level(if conf.deps_debug {
            LevelFilter::Debug
        } else {
            LevelFilter::Warn
        })
        .with_module_level(
            "teos",
            if conf.debug {
                LevelFilter::Debug
            } else {
                LevelFilter::Info
            },
        )
        .init()
        .unwrap();

    // log default data dir
    log::info!("Using default data directory: {:?}", &path);

    // Log config file path
    log::info!("Using configuration file: {:?}", &conf_file_path);`

Reply:
> > 
> 
> I referenced it with `let mut conf = config::from_file::<Config>(&conf_file_path);` but still shows error and states `consider removing the borrow`.

That doesn't make sense. Have you updated the method signature? Just tested it and it works.

```diff
diff --git a/teos/src/cli.rs b/teos/src/cli.rs
index 5c20f00..eba032b 100644
--- a/teos/src/cli.rs
+++ b/teos/src/cli.rs
@@ -27,7 +27,7 @@ async fn main() {
     let command = opt.command.clone();

     // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
-    let mut conf = config::from_file::<Config>(path.join("teos.toml"));
+    let mut conf = config::from_file::<Config>(&path.join("teos.toml"));
     conf.patch_with_options(opt);

     let key = fs::read(&path.join("client-key.pem"))
diff --git a/teos/src/config.rs b/teos/src/config.rs
index 1669b25..50fb8a3 100644
--- a/teos/src/config.rs
+++ b/teos/src/config.rs
@@ -16,7 +16,7 @@ pub fn data_dir_absolute_path(data_dir: String) -> PathBuf {
     }
 }

-pub fn from_file<T: Default + serde::de::DeserializeOwned>(path: PathBuf) -> T {
+pub fn from_file<T: Default + serde::de::DeserializeOwned>(path: &PathBuf) -> T {
     match std::fs::read(path) {
         Ok(file_content) => toml::from_slice::<T>(&file_content).map_or_else(
             |e| {
diff --git a/teos/src/main.rs b/teos/src/main.rs
index e7b8f37..1ab249e 100644
--- a/teos/src/main.rs
+++ b/teos/src/main.rs
@@ -71,9 +71,6 @@ async fn main() {
     let path = config::data_dir_absolute_path(opt.data_dir.clone());
     let conf_file_path = path.join("teos.toml");

-    // Log config file path
-    log::info!("Using configuration file: {:?}", conf_file_path);
-
     // Create data dir if it does not exist
     fs::create_dir_all(&path).unwrap_or_else(|e| {
         eprintln!("Cannot create data dir: {e:?}");
@@ -81,7 +78,7 @@ async fn main() {
     });

     // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
-    let mut conf = config::from_file::<Config>(conf_file_path);
+    let mut conf = config::from_file::<Config>(&conf_file_path);
     let is_default = conf.is_default();
     conf.patch_with_options(opt);
     conf.verify().unwrap_or_else(|e| {
@@ -89,9 +86,6 @@ async fn main() {
         std::process::exit(1);
     });

-    // log default data dir
-    log::info!("Using default data directory: {:?}", &path);
-
     // log the non-default options that the daemon is using
     if !is_default {
         if conf.api_bind != "localhost" {
@@ -165,6 +159,12 @@ async fn main() {
         log::info!("Loading configuration from file")
     }

+    // Log config file path
+    log::info!("Using configuration file: {:?}", conf_file_path);
+
+    // log default data dir
+    log::info!("Using default data directory: {:?}", &path);
+
     // Create network dir
     let path_network = path.join(conf.btc_network.clone());
     fs::create_dir_all(&path_network).unwrap_or_else(|e| {
```

Reply:
> > > 
> > 
> > 
> > I referenced it with `let mut conf = config::from_file::<Config>(&conf_file_path);` but still shows error and states `consider removing the borrow`.
> 
> That doesn't make sense. Have you updated the method signature? Just tested it and it works.
> 
> ```diff
> diff --git a/teos/src/cli.rs b/teos/src/cli.rs
> index 5c20f00..eba032b 100644
> --- a/teos/src/cli.rs
> +++ b/teos/src/cli.rs
> @@ -27,7 +27,7 @@ async fn main() {
>      let command = opt.command.clone();
> 
>      // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
> -    let mut conf = config::from_file::<Config>(path.join("teos.toml"));
> +    let mut conf = config::from_file::<Config>(&path.join("teos.toml"));
>      conf.patch_with_options(opt);
> 
>      let key = fs::read(&path.join("client-key.pem"))
> diff --git a/teos/src/config.rs b/teos/src/config.rs
> index 1669b25..50fb8a3 100644
> --- a/teos/src/config.rs
> +++ b/teos/src/config.rs
> @@ -16,7 +16,7 @@ pub fn data_dir_absolute_path(data_dir: String) -> PathBuf {
>      }
>  }
> 
> -pub fn from_file<T: Default + serde::de::DeserializeOwned>(path: PathBuf) -> T {
> +pub fn from_file<T: Default + serde::de::DeserializeOwned>(path: &PathBuf) -> T {
>      match std::fs::read(path) {
>          Ok(file_content) => toml::from_slice::<T>(&file_content).map_or_else(
>              |e| {
> diff --git a/teos/src/main.rs b/teos/src/main.rs
> index e7b8f37..1ab249e 100644
> --- a/teos/src/main.rs
> +++ b/teos/src/main.rs
> @@ -71,9 +71,6 @@ async fn main() {
>      let path = config::data_dir_absolute_path(opt.data_dir.clone());
>      let conf_file_path = path.join("teos.toml");
> 
> -    // Log config file path
> -    log::info!("Using configuration file: {:?}", conf_file_path);
> -
>      // Create data dir if it does not exist
>      fs::create_dir_all(&path).unwrap_or_else(|e| {
>          eprintln!("Cannot create data dir: {e:?}");
> @@ -81,7 +78,7 @@ async fn main() {
>      });
> 
>      // Load conf (from file or defaults) and patch it with the command line parameters received (if any)
> -    let mut conf = config::from_file::<Config>(conf_file_path);
> +    let mut conf = config::from_file::<Config>(&conf_file_path);
>      let is_default = conf.is_default();
>      conf.patch_with_options(opt);
>      conf.verify().unwrap_or_else(|e| {
> @@ -89,9 +86,6 @@ async fn main() {
>          std::process::exit(1);
>      });
> 
> -    // log default data dir
> -    log::info!("Using default data directory: {:?}", &path);
> -
>      // log the non-default options that the daemon is using
>      if !is_default {
>          if conf.api_bind != "localhost" {
> @@ -165,6 +159,12 @@ async fn main() {
>          log::info!("Loading configuration from file")
>      }
> 
> +    // Log config file path
> +    log::info!("Using configuration file: {:?}", conf_file_path);
> +
> +    // log default data dir
> +    log::info!("Using default data directory: {:?}", &path);
> +
>      // Create network dir
>      let path_network = path.join(conf.btc_network.clone());
>      fs::create_dir_all(&path_network).unwrap_or_else(|e| {
> ```

Yes, working now. Forgot to update the borrow of cli.rs file. Thank you!